rewrite the Concepts section

erig0 opened the pull request:
I tried to keep the new Concepts page brief. I want to highlight the most important concepts without overwhelming the reader.

TorontoMedia commented on documentation/concepts.md:
````diff
@@ -3,17 +3,114 @@ layout: page
 
-More information on the firewalld D-Bus API is available [here](man-pages/firewalld.dbus.html).
+```
++----------+     policyA     +----------+
````
It appears that the concept is clearly outlined and is easily understood. How do you feel about adding styled graphs with the jekyll-mermaid plugin?
```mermaid
graph LR;
    A("internal<br>(zone)") -->|policyB| B("external<br>(zone)")
    B -->|policyA| A 
```

erig0 replied:
Is the plugin enable by default?

erig0 replied:
I found this blog post: https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/

But it's unclear to me if it's enabled in github pages, or just repo markdown files.

erig0 replied:
I'll use the mermaid, if it doesn't render we'll figure it out after the merge. ;)

TorontoMedia replied:
> So it definitely took some playing around to get it to work. I ended up having to add some script tags to the `page` layout. I will probably tweak some settings.

It looks great, thanks!

TorontoMedia commented on documentation/concepts.md:
````diff
@@ -3,17 +3,114 @@ layout: page
+zone.
+
+```
++----------+                 +----------+
````
```mermaid
graph LR;
    A("internal<br>(zone)") -->|myPolicy| B("external<br>(zone)")
```

TorontoMedia commented:
There are very few references to the connection between Network-Manager and FirewallD. Should the reference be merely clarified rather than entirely removed?

n0-1 commented on documentation/concepts.md:
```diff
@@ -0,0 +1,116 @@
+
+## Runtime and Permanent Configuration
+
+Firewalld maintains separate runtime and permanent configurations. This
+allows runtime only changes. Manipulating the permanent configuration
+requires using the `--permanent` CLI option.
```
Enhance this by: "Runtime configuration does not persist during program restarts (or machine reboots). At startup it is populated from persistent configuration." (In better English probably. ;)

erig0 replied:
Reworded:

```
Firewalld maintains separate runtime and permanent configurations. This
allows runtime only changes. The runtime configuration does not persist
after firewalld reload or restart. At startup it is populated from the
permanent configuration. Manipulating the permanent configuration
requires using the `--permanent` option.
```

erig0 commented:
> There are very few references to the connection between NetworkManager and FirewallD. Should the reference be merely clarified rather than entirely removed?

I don't think it's worth covering in the concepts. It's better left to a dedicate doc page.

erig0 commented on documentation/concepts.md:
```diff
@@ -0,0 +1,116 @@
+
+Policies may apply firewall rules to a variety of traffic, e.g.
+forwarded traffic. This is useful for network level filtering.
+
```
That's correct. Attaching rules to zones is mostly historical. Ideally they'd just be containers to attach interface/sources.

Should I make it clear that zones are internally implemented via policies?

Should I de-emphasis attaching rules to zones?

erig0 replied:
Is this better or worse? :D

```
Policies may apply firewall rules to all traffic types. This includes
filtering input, forwarded, and output traffic; as well as performing
network address translations (NAT).

In addition to logically separating the network, zones implicitly
provide a policy for traffic destined to the host running firewalld
(input).
```

erig0 commented on documentation/concepts.md:
```diff
@@ -0,0 +1,116 @@
+
+Principle 4 and 5 are actually a consequence of Principle 3.
+
+Principle 4 is configurable via CLI option `--remove-forward`.
```
I'll change it to "allowed by default".

erig0 replied:
New text: 

```
Firewalld follows some strict principles in regards to zones.

1. traffic ingresses one and only one zone
2. traffic egresses one and only one zone
3. a zone defines a level of trust
4. intra-zone (with-in the same zone) is allowed by default
5. inter-zone (zone to zone) is denied by default

Principle 4 and 5 are actually a consequence of Principle 3.

Principle 4 is configurable via zone option `--remove-forward`.
Principle 5 is configurable by adding new policies.
```

erig0 commented on documentation/concepts.md:
```diff
@@ -0,0 +1,116 @@
+Principle 4 is configurable via CLI option `--remove-forward`.
+
+## End Station Firewall
+
```
Nope. Just trying to show use cases.

erig0 replied:
I re-titled these sections.
```
## Simple Use Case: End Station Firewall
```

```
## Complex Use Case: Network Firewall
```

erig0 commented on documentation/concepts.md:
```diff
@@ -0,0 +1,116 @@
+
+## Network Firewall
+
+Policies apply firewall rules in a stateful, unidirectional manner. This
+means you only need to consider one direction of the traffic. The return
+traffic is implicitly allowed.
```
I don't think so. The policy filters traffic in one direction. The return traffic is allowed due to statefulness.

erig0 replied:
Does s/apply firewall rules/filter traffic/ make it clear?
i.e.
```
Policies filtering traffic in a stateful, unidirectional manner. This
means you only need to consider one direction of the traffic. The return
traffic is implicitly allowed.
```

erig0 replied:
Reworded a bit. I hope it helps.
```
Policies apply firewall rules in a stateful, unidirectional manner. This
means you only need to consider one direction of the traffic. The
traffic's return path is implicitly allowed due to firewalld's stateful
filtering.
```

erig0 commented on documentation/concepts.md:
````diff
@@ -0,0 +1,116 @@
++----------+                 +----------+
+```
+
+This is a common scenario for a home network. To allow the LAN
+(internal) access the internet you can allow all traffic by setting the
+default target of the policy using `--set-target`.
````
This example uses the `external` zone, which has masquerade enabled by default.

Should I add an example to explicitly enable masquerade?

erig0 commented:
Hi all, I pushed an update that I think addresses all your feedback. I rearranged things quite a bit, so please do another full review. Thanks in advance!

TorontoMedia commented on documentation/concepts.md:
```diff
@@ -0,0 +1,116 @@
+
+## Introduction
+
+firewalld aims to be a simple, zoning, stateful firewall. Firewall rules
+are organized into zones and policies. Zones logically separate the
+network. Policies apply rules to traffic flowing between zones.
```
```suggestion
Firewalld is a simple, stateful, zoned based firewall with a number of features, including policies. Policies and zones are used to organize firewall rules. The network is logically divided into zones and the traffic flowing between them can be governed by policies.
```

TorontoMedia replied:
I added a suggestion to make the introductory summary a little less terse.

erig0 replied:
Good suggestion. I dropped part of the first sentence.

```
Firewalld is a simple, stateful, zoned based firewall. Policies and
zones are used to organize firewall rules. The network is logically
divided into zones and the traffic flowing between them can be governed
by policies.
```

ben-hart-jamf commented:
Looks good to me!

kpfleming commented on documentation/concepts.md:
```diff
@@ -3,17 +3,115 @@ layout: page
 
-![firewalld-structure+nftables](firewalld-structure+nftables.png "firewalld structure")
-*firewalld structure*
+Firewalld is a simple, stateful, zoned based firewall. Policies and
```
Maybe 'zone-based' instead of 'zoned based'.

kpfleming commented on documentation/concepts.md:
````diff
@@ -3,17 +3,115 @@ layout: page
+rejected.
+
+```
+# firewall-cmd --policy myPolicy --add-service http
````
I think this is where users get lost... there is no indication in this command that any zones are being affected. Does this command 'work' because the firewalld default zone on the machine on the strict network is 'internal'?

erig0 replied:
No. This command is affecting a policy, not a zone. Note the `--policy` option.

kpfleming replied:
So then what causes `myPolicy` to be applied to any zones in particular? I seem to have missed that.

erig0 replied:
In this example `myPolicy` affects traffic forwarded the `internal` zone to the `external` zone (explained in lines 51-53). It does not "apply to a zone" directly.

https://github.com/firewalld/firewalld.github.io/pull/25/commits/f68ad2461f121e561564a131dc05f0ff8a93d629#diff-296328860e666fb622185f2b4951481a189a5f689bbbea8cc2d7eabd0d50a057R51-R53

kpfleming replied:
OK, that's helpful. I guess my remaining concern is learning how that configuration occurred... should I assume that some command *before* the one shown created `myPolicy` and connected it to those zones? If so, then I'm happy :-)

erig0 replied:
Yes. I didn't show the creation of `myPolicy` because I was trying to show only the concepts. Would it help if I showed the CLI output for `myPolicy`? It's below.

```
[root@vm-bos-rhel8 ~]# firewall-cmd --info-policy myPolicy
myPolicy (active)
  priority: -1
  target: CONTINUE
  ingress-zones: internal
  egress-zones: external
  services: 
  ports: 
  protocols: 
  masquerade: no
  forward-ports: 
  source-ports: 
  icmp-blocks: 
  rich rules: 
```

And these are the commands that create `myPolicy`:
```
# firewall-cmd --permanent --new-policy myPolicy
# firewall-cmd --permanent --policy myPolicy --add-ingress-zone internal
# firewall-cmd --permanent --policy myPolicy --add-egress-zone external
```

kpfleming replied:
In my opinion that would help quite a lot :-)

erig0 replied:
I pushed the update. Please take a look.

kpfleming reviewed (commented):
Overall this is really helpful!

erig0 commented:
Thanks for the collaboration everyone! Merging.
